### PR TITLE
Public notebooks for causal graph testing

### PR DESCRIPTION
**Pull request by timothyb0912**
Branch for adding the public notebook demonstrating testing of one's causal graph.

### CODE REVIEW

**Comment by timothyb0912** [comment]
@bouzaghrane and @hassanobeid1994 
Feel free to let me know if you have any suggestions / comments / questions about the notebook skeleton that I uploaded. I'll fill it in over the next couple of days.

**Comment by bouzaghrane** [comment]
The setup looks good to me!

**Comment by timothyb0912** [comment]
@hassanobeid1994 : Please take a look! I've finished the first draft of the public notebook.
cc @bouzaghrane for visibility.

File: notebooks/final/_04-tb-testing-your-causal-graph.md [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+    seed=1038,
+    num_permutations=NUM_PERMUTATIONS,
+    permutation_color=PERMUTED_COLOR)
+```
+
+### Caveats and pitfalls
+Two important issues when testing the assumption of mean independence between $X_1$ and $X_2$ are underfitting and overfitting of the model for $E \left[ X_1 \mid X_2 \right]$.
+
+If $E \left[ X_1 \mid X_2 \right]$ is underfit, then one's observed test statitic ($r^2$) will be lower than it would be under a correctly specified model.
+This will increase the probability of one failing to reject the null-hypothesis when the null is false.
```
Comment (hassan-obeid) [inline notebooks/final/_04-tb-testing-your-causal-graph.md]: add: ", or type II error" for consistency.

File: notebooks/final/_04-tb-testing-your-causal-graph.md [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+oi.visual_permutation_test(
+    license_array,
+    num_cars_array,
+    z_array=None,
+    seed=1038,
+    num_permutations=NUM_PERMUTATIONS,
+    permutation_color=PERMUTED_COLOR)
+```
+
+### Caveats and pitfalls
```
Comment (hassan-obeid) [inline notebooks/final/_04-tb-testing-your-causal-graph.md:164]: Not urgent, but is it worth it to split this into caveats/pitfalls and how to overcome them? The write-up already touches on how to avoid underfitting/overfitting, so it may be worth it to split this section into "problems" and "solutions" kind of format.

File: notebooks/final/_04-tb-testing-your-causal-graph.md [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+If the conditional mean of $X_1$ depends on un-modeled functions of $Z$ and $Z \rightarrow X_2$, then the inclusion of $X_2$ in one's model may serve as a proxy for the un-modeled function of $Z$.
+Such proxy behavior would lead one to observe "inflated" values of $r^2$ when modeling $E \left[ X_1 \mid X_2, Z \right]$, thus increasing the probability that one will falsely reject the null-hypothesis of conditional independence.
+
+In other words, to guard against higher-than-nominal probabilities of type-1 error, one needs to guard against underfitting of the model for $E \left[ X_1 \mid Z \right]$ **before** computing one's test-statistic and (permutation-based) reference-distribution.
+
+The second issue is misspecification of $E \left[ X_1 \mid X_2, Z \right]$.
+If one's model for $E \left[ X_1 \mid X_2, Z \right]$ is underfit with respect to $X_2$, then one's test-statistic ($r^2$) will be lower than it should be under accurate specification.
+This leads to lower probability of correctly rejecting the null-hypothesis.
+
+The last two issues are the converse of the problems and effects above.
```
Comment (hassan-obeid) [inline notebooks/final/_04-tb-testing-your-causal-graph.md]: Which above? The marginal independence tests?

File: notebooks/final/_04-tb-testing-your-causal-graph.md [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+If one's model for $E \left[ X_1 \mid Z \right]$ is overfit, then the reference-distribution of the test statistic will be artificially high.
```
Comment (hassan-obeid) [inline notebooks/final/_04-tb-testing-your-causal-graph.md:228]: Is it a reference distribution or a reference value?

Comment (hassanobeid1994) [inline notebooks/final/_04-tb-testing-your-causal-graph.md:228]: Is it a reference distribution or a reference value?

File: notebooks/final/_04-tb-testing-your-causal-graph.md [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+This will lead to lower probability of correctly rejecting the null hypothesis of conditional independence.
```
Comment (hassan-obeid) [inline notebooks/final/_04-tb-testing-your-causal-graph.md]: I suggest we rephrase this to standard hypothesis testing lingo. Instead of "lower the probability of correctly rejecting the null hypothesis", I would use "this will increase the probability of failing to reject the null hypothesis when it's true".

**Review by hassan-obeid — CHANGES_REQUESTED** [review]
review in-progress -- will add more comments soon.

File: notebooks/final/_04-tb-testing-your-causal-graph.md (hunk 4 above)
Comment (timothyb0912) [reply notebooks/final/_04-tb-testing-your-causal-graph.md:228]: I literally mean the reference-distribution, as written, in this sentence. I'm referring to the permutation-based distribution that one compares the test-statistic to.

File: notebooks/final/_04-tb-testing-your-causal-graph.md (hunk 1 above)
Comment (timothyb0912) [reply notebooks/final/_04-tb-testing-your-causal-graph.md]: Good call!

File: notebooks/final/_04-tb-testing-your-causal-graph.md (hunk 3 above)
Comment (timothyb0912) [reply notebooks/final/_04-tb-testing-your-causal-graph.md]: Will clarify! Thanks for the call-out!

File: notebooks/final/_04-tb-testing-your-causal-graph.md (hunk 5 above)
Comment (timothyb0912) [reply notebooks/final/_04-tb-testing-your-causal-graph.md]: Just addressed (although not using the exact wording you suggested).

**Review by hassan-obeid — APPROVED** [review]
